<a href="https://colab.research.google.com/github/seecode4/seeRepo1/blob/main/mec2-projects/Student_MLE_MiniProject_Recommendation_Engines.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Mini Project: Recommendation Engines

Recommendation engines are algorithms designed to provide personalized suggestions or recommendations to users. These systems analyze user behavior, preferences, and interactions with items (products, movies, music, articles, etc.) to predict and offer items that users are likely to be interested in. Recommendation engines play a crucial role in enhancing user experience, driving engagement, and increasing conversion rates in various applications, including e-commerce, entertainment, content platforms, and more.

There are generally two approaches taken in collaborative filtering and content-based recommendation engines:

**1. Collaborative Filtering:**
Collaborative Filtering is a popular approach to building recommendation systems that leverages the collective behavior of users to make personalized recommendations. It is based on the idea that users who have agreed in the past will likely agree in the future. There are two main types of collaborative filtering:

- **User-based Collaborative Filtering:** This method finds users similar to the target user based on their past interactions (e.g., ratings or purchases). It then recommends items that similar users have liked but the target user has not interacted with yet.

- **Item-based Collaborative Filtering:** In this approach, the system identifies similar items based on user interactions. It recommends items that are similar to the ones the target user has already liked or interacted with.

Collaborative filtering does not require any explicit information about items but relies on the similarity between users or items. It is effective in capturing complex patterns and can provide serendipitous recommendations. However, it suffers from the cold-start problem (i.e., difficulty in recommending to new users or items with no interactions) and scalability challenges in large datasets.

**2. Content-Based Recommendation:**
Content-based recommendation is an alternative approach to building recommendation systems that focuses on the attributes or features of items and users. It leverages the characteristics of items to make recommendations. The key steps involved in content-based recommendation are:

- **Feature Extraction:** For each item, relevant features are extracted. For movies, these features could be genre, director, actors, and plot summary.

- **User Profile:** A user profile is created based on the items they have interacted with in the past. The user profile contains the weighted importance of features based on their interactions.

- **Similarity Calculation:** The similarity between items or between items and the user profile is calculated using similarity metrics like cosine similarity or Euclidean distance.

- **Recommendation:** Items that are most similar to the user profile are recommended to the user.

Content-based recommendation systems are less affected by the cold-start problem as they can still recommend items based on their features. They are also more interpretable as they rely on item attributes. However, they may miss out on providing serendipitous recommendations and can be limited by the quality of feature extraction and user profiles.

**Choosing Between Collaborative Filtering and Content-Based:**
Both collaborative filtering and content-based approaches have their strengths and weaknesses. The choice between them depends on the specific requirements of the recommendation system, the type of data available, and the user base. Hybrid approaches that combine collaborative filtering and content-based techniques are also common, aiming to leverage the strengths of both methods and mitigate their weaknesses.

In this mini-project, you'll be building both content based and collaborative filtering engines for the [MovieLens 25M dataset](https://grouplens.org/datasets/movielens/25m/). The MovieLens 25M dataset is one of the most widely used and popular datasets for building and evaluating recommendation systems. It is provided by the GroupLens Research project, which collects and studies datasets related to movie ratings and recommendations. The MovieLens 25M dataset contains movie ratings and other related information contributed by users of the MovieLens website.

**Dataset Details:**
- **Size:** The dataset contains approximately 25 million movie ratings.
- **Users:** It includes ratings from over 162,000 users.
- **Movies:** The dataset consists of ratings for more than 62,000 movies.
- **Ratings:** The ratings are provided on a scale of 1 to 5, where 1 is the lowest rating and 5 is the highest.
- **Timestamps:** Each rating is associated with a timestamp, indicating when the rating was given.

**Data Files:**
The dataset is usually split into three CSV files:

1. **movies.csv:** Contains information about movies, including the movie ID, title, genres, and release year.
   - Columns: movieId, title, genres

2. **ratings.csv:** Contains movie ratings provided by users, including the user ID, movie ID, rating, and timestamp.
   - Columns: userId, movieId, rating, timestamp

3. **tags.csv:** Contains user-generated tags for movies, including the user ID, movie ID, tag, and timestamp.
   - Columns: userId, movieId, tag, timestamp

First, import all the libraries you'll need.

In [1]:
import zipfile
import numpy as np
import pandas as pd
from urllib.request import urlretrieve
from sklearn.metrics.pairwise import cosine_similarity

Next, download the relevant components of the MoveLens dataset. Note, these instructions are roughly based on the colab [here](https://colab.research.google.com/github/google/eng-edu/blob/main/ml/recommendation-systems/recommendation-systems.ipynb?utm_source=ss-recommendation-systems&utm_campaign=colab-external&utm_medium=referral&utm_content=recommendation-systems#scrollTo=O3bcgduFo4s6).

In [2]:
print("Downloading movielens data...")

urlretrieve('http://files.grouplens.org/datasets/movielens/ml-100k.zip', 'movielens.zip')
zip_ref = zipfile.ZipFile('movielens.zip', 'r')
zip_ref.extractall()
print("Done. Dataset contains:")
print(zip_ref.read('ml-100k/u.info'))

ratings_cols = ['user_id', 'movie_id', 'rating', 'unix_timestamp']
ratings = pd.read_csv(
    'ml-100k/u.data', sep='\t', names=ratings_cols, encoding='latin-1')

# The movies file contains a binary feature for each genre.
genre_cols = [
    "genre_unknown", "Action", "Adventure", "Animation", "Children", "Comedy",
    "Crime", "Documentary", "Drama", "Fantasy", "Film-Noir", "Horror",
    "Musical", "Mystery", "Romance", "Sci-Fi", "Thriller", "War", "Western"
]
movies_cols = [
    'movie_id', 'title', 'release_date', "video_release_date", "imdb_url"
] + genre_cols
movies = pd.read_csv(
    'ml-100k/u.item', sep='|', names=movies_cols, encoding='latin-1')

Done. Dataset contains:
b'943 users\n1682 items\n100000 ratings\n'


Before doing any kind of machine learning, it's always good to familiarize yourself with the datasets you'lll be working with.

Here are your tasks:

1. Spend some time familiarizing yourself with both the `movies` and `ratings` dataframes. How many unique user ids are present? How many unique movies are there?
2. Create a new dataframe that merges the `movies` and `ratings` tables on 'movie_id'. Only keep the 'user_id', 'title', 'rating' fields in this new dataframe.

In [3]:
# Spend some time familiarizing yourself with both the movies and ratings
# dataframes. How many unique user ids are present? How many unique movies
# are there?
# ratings_cols = ['user_id', 'movie_id', 'rating', 'unix_timestamp']
# genre_cols =  ['genre_unknown', 'Action', 'Adventure', 'Animation', 'Children', 'Comedy', 'Crime', 'Documentary', 'Drama', 'Fantasy', 'Film-Noir', 'Horror', 'Musical', 'Mystery', 'Romance', 'Sci-Fi', 'Thriller', 'War', 'Western']
# movies_cols = ['movie_id', 'title', 'release_date', "video_release_date", "imdb_url"] + genre_cols
print(ratings.head())
print(movies.head())
print('-'*25)
print("ratings shape:", ratings.shape)
print("movies.shape:", movies.shape)
print("ratings - num unique user_id:", len(ratings['user_id'].unique()))
print("ratings - num unique movie_id:", len(ratings['movie_id'].unique()))
print("movies - num unique movie_id:", len(movies['movie_id'].unique()))
print("num_genres:", len(genre_cols))

   user_id  movie_id  rating  unix_timestamp
0      196       242       3       881250949
1      186       302       3       891717742
2       22       377       1       878887116
3      244        51       2       880606923
4      166       346       1       886397596
   movie_id              title release_date  video_release_date  \
0         1   Toy Story (1995)  01-Jan-1995                 NaN   
1         2   GoldenEye (1995)  01-Jan-1995                 NaN   
2         3  Four Rooms (1995)  01-Jan-1995                 NaN   
3         4  Get Shorty (1995)  01-Jan-1995                 NaN   
4         5     Copycat (1995)  01-Jan-1995                 NaN   

                                            imdb_url  genre_unknown  Action  \
0  http://us.imdb.com/M/title-exact?Toy%20Story%2...              0       0   
1  http://us.imdb.com/M/title-exact?GoldenEye%20(...              0       1   
2  http://us.imdb.com/M/title-exact?Four%20Rooms%...              0       0   
3  http://u

In [4]:
def genres_str(df_row):
  gstr = ''
  for i in range(len(genre_cols)):
    gname = genre_cols[i]
    if df_row[gname]:
      gstr += ' ' + gname
  return gstr.strip()

movies['genres'] = ''
print(type(movies.iloc[0,:]))
movies['genres'] = movies.apply(lambda x: genres_str(x), axis=1)
print(movies.iloc[0:3,-1])

<class 'pandas.core.series.Series'>
0    Animation Children Comedy
1    Action Adventure Thriller
2                     Thriller
Name: genres, dtype: object


In [5]:
# Merge movies and ratings dataframes
df_all_movie_info = pd.merge(movies, ratings, on='movie_id', how='inner')
print(f'df_all_movie_info:\n shape={df_all_movie_info.shape}',
      f'\n columns={df_all_movie_info.columns}')
# print(df_all_movie_info.head(3))

# filter for columns of interest
df_movie_ratings = df_all_movie_info[['user_id', 'movie_id', 'title', 'rating', 'genres']]
# df_movie_ratings.set_index(['user_id', 'movie_id'], inplace=True)
print(f'df_movie_ratings:\n shape={df_movie_ratings.shape}',
      f'\n columns={df_movie_ratings.columns}')
print(df_movie_ratings.head())
print('-----------')
print(df_movie_ratings['title'].value_counts())
print(df_movie_ratings[df_movie_ratings['title'] == 'Star Wars (1977)'].head())

df_all_movie_info:
 shape=(100000, 28) 
 columns=Index(['movie_id', 'title', 'release_date', 'video_release_date', 'imdb_url',
       'genre_unknown', 'Action', 'Adventure', 'Animation', 'Children',
       'Comedy', 'Crime', 'Documentary', 'Drama', 'Fantasy', 'Film-Noir',
       'Horror', 'Musical', 'Mystery', 'Romance', 'Sci-Fi', 'Thriller', 'War',
       'Western', 'genres', 'user_id', 'rating', 'unix_timestamp'],
      dtype='object')
df_movie_ratings:
 shape=(100000, 5) 
 columns=Index(['user_id', 'movie_id', 'title', 'rating', 'genres'], dtype='object')
   user_id  movie_id             title  rating                     genres
0      308         1  Toy Story (1995)       4  Animation Children Comedy
1      287         1  Toy Story (1995)       5  Animation Children Comedy
2      148         1  Toy Story (1995)       4  Animation Children Comedy
3      280         1  Toy Story (1995)       4  Animation Children Comedy
4       66         1  Toy Story (1995)       3  Animation Childre

As mentioned in the introduction, content-Based Filtering is a recommendation engine approach that focuses on the attributes or features of items (products, movies, music, articles, etc.) and leverages these features to make personalized recommendations. The underlying idea is to match the characteristics of items with the preferences of users to suggest items that align with their interests. Content-based filtering is particularly useful when explicit user-item interactions (e.g., ratings or purchases) are sparse or unavailable.

**Key Steps in Content-Based Filtering:**

1. **Feature Extraction:**
   - For each item, relevant features are extracted. These features are typically descriptive attributes that can be represented numerically, such as genre, director, actors, author, publication date, and keywords.
   - In the case of text-based items, natural language processing techniques may be used to extract features like TF-IDF (Term Frequency-Inverse Document Frequency) scores.

2. **User Profile Creation:**
   - A user profile is created based on the items they have interacted with in the past. The user profile contains the weighted importance of features based on their interactions.
   - For example, if a user has watched several action movies, the action genre feature would receive a higher weight in their profile.

3. **Similarity Calculation:**
   - The similarity between items or between items and the user profile is calculated using similarity metrics like cosine similarity, Euclidean distance, or Pearson correlation.
   - Cosine similarity is commonly used as it measures the cosine of the angle between two vectors, which represents their similarity.

4. **Recommendation:**
   - Items that are most similar to the user profile are recommended to the user. These are items whose features have the highest similarity scores with the user profile.
   - The recommended items are presented as a list sorted by their similarity scores.

**Advantages of Content-Based Filtering:**
1. **No Cold-Start Problem:** Content-based filtering can make recommendations even for new users with no historical interactions because it relies on item features rather than user history.

2. **User Independence:** The recommendations are based solely on the features of items and do not require knowledge of other users' preferences or behavior.

3. **Transparency:** Content-based recommendations are interpretable, as they depend on the features of items, making it easier for users to understand why specific items are recommended.

4. **Serendipity:** Content-based filtering can recommend items with characteristics not seen before by the user, leading to serendipitous discoveries.

5. **Diversity in Recommendations:** The method can offer diverse recommendations since it suggests items with different feature combinations.

**Limitations of Content-Based Filtering:**
1. **Limited Discovery:** Content-based filtering may struggle to recommend items outside the scope of users' historical interactions or interests.

2. **Over-Specialization:** Users may receive recommendations that are too similar to their previous choices, leading to a lack of exposure to new item categories.

3. **Dependency on Feature Quality:** The quality and relevance of item features significantly influence the quality of recommendations.

4. **Limited for Cold Items:** Content-based filtering can struggle to recommend new items with limited feature information.

Here is your task:

1. Write a function that takes in a user id and the dataframe you created before that contains 'user_id', 'title', and 'rating'. The function should return content-based recommendations for this user. Here are steps you can take:

  A. Get the user's rated movies

  B. Create a TF-IDF matrix using movie genres. Note, this can be extracted from the `movies` dataframe.

  C. Compute the cosine similarity between movie genres. Use the [cosine_similarity](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.pairwise.cosine_similarity.html) function.

  D. Get the indices of similar movies to those rated by the user based on cosine similarity. Keep only the top 5.

  E. Remove duplicates and movies already rated by the user.

In [6]:
# Content-Based Filtering using Movie Genres

from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity

# ratings_cols = ['user_id', 'movie_id', 'rating', 'unix_timestamp']
# genre_cols =  ['genre_unknown', 'Action', 'Adventure', 'Animation', 'Children', 'Comedy', 'Crime', 'Documentary', 'Drama', 'Fantasy', 'Film-Noir', 'Horror', 'Musical', 'Mystery', 'Romance', 'Sci-Fi', 'Thriller', 'War', 'Western']
# movies_cols = ['movie_id', 'title', 'release_date', "video_release_date", "imdb_url"] + genre_cols
# df_movie_ratings
def content_based_recommendation(user_id, df):
  # Get the user's rated movies
  df_user_movie_ratings = df[df['user_id'] == user_id].sort_values(by='title')
  # print("\n df_user_movie_ratings shape:", df_user_movie_ratings.shape)
  # print("\n df_user_movie_ratings index:\n", df_user_movie_ratings.index)

  # Create a TF-IDF matrix using movie genre
  tfidf = TfidfVectorizer(max_df=0.7, min_df=1)
  tfidf_matrix = tfidf.fit_transform(movies['genres'])
  # print('tfidf_matrix:\n', tfidf_matrix.shape, tfidf_matrix)

  # a row for each movie, a column for each genre
  tfidf_df = pd.DataFrame(tfidf_matrix.toarray(),
                          columns=tfidf.get_feature_names_out())
  tfidf_df.index = movies['title']
  # print("tfidf_df:\n", type(tfidf_df), tfidf_df.shape, "\n",
  #       tfidf_df.head(2), "\n tfidf_df.index[0:10]", tfidf_df.index[0:10])

  # Compute the cosine similarity between movie genres
  cosine_similarity_array = cosine_similarity(tfidf_df)
  df_genre_similarities = pd.DataFrame(cosine_similarity_array,
                                       index=tfidf_df.index,
                                       columns=tfidf_df.index)

  # Get the indices of the similar movies to those rated by the user
  list_of_movies_seen = list(df_user_movie_ratings['title'])

  user_movies_df = tfidf_df.loc[list_of_movies_seen]
  # print("user_movies_df:\n", type(user_movies_df), "len=", len(user_movies_df))
  # print(user_movies_df)

  # subset of movies not seen/rated by user
  non_user_movies_df = tfidf_df.drop(list_of_movies_seen, axis=0)
  # print("non_user_movies_df:\n", type(non_user_movies_df), "len=", len(non_user_movies_df))

  # based on cosine similarity; keep only the top 5
  user_prof = user_movies_df.mean()
  # find movies matching user_prof that user has not seen
  similarity_array = cosine_similarity(user_prof.values.reshape(1, -1), non_user_movies_df)
  similarity_df = pd.DataFrame(similarity_array.T,
                               index=non_user_movies_df.index,
                               columns=["similarity_score"])
  sorted_similarity_df = similarity_df.sort_values(by="similarity_score",
                                                   ascending=False)
  # print(sorted_similarity_df.shape, sorted_similarity_df.head(8))
  # print("\n\nTop 5 similar movies:\n", sorted_similarity_df.iloc[0:5,:])
  # Remove duplicates and movies already rated by the user
  return sorted_similarity_df.iloc[0:5,:]

user_ids = [1, 8, 148, 166, 186, 196, 244, 308]
for id in user_ids:
  top_movies = content_based_recommendation(id, df_movie_ratings)
  print(f"\nTop 5 (content_based) similar movies for user_id {id} \n", top_movies)
  print('--------------------')


Top 5 (content_based) similar movies for user_id 1 
                                                     similarity_score
title                                                               
Night on Earth (1991)                                       0.731446
Paper, The (1994)                                           0.731446
To Die For (1995)                                           0.731446
Tin Men (1987)                                              0.731446
Adventures of Priscilla, Queen of the Desert, T...          0.731446
--------------------

Top 5 (content_based) similar movies for user_id 8 
                                        similarity_score
title                                                  
Lost World: Jurassic Park, The (1997)          0.788149
Abyss, The (1989)                              0.788149
Escape from L.A. (1996)                        0.788149
Escape from New York (1981)                    0.788149
Men in Black (1997)                            0.758

The key idea behind collaborative filtering is that users who have agreed in the past will likely agree in the future. Instead of relying on item attributes or user profiles, collaborative filtering identifies patterns of user behavior and item preferences from the interactions present in the data.

**Types of Collaborative Filtering:**
There are two main types of collaborative filtering:

**Collaborative Filtering Process:**
The collaborative filtering process typically involves the following steps:

1. **Data Collection:**
   - Gather data on user-item interactions, such as movie ratings, product purchases, or article clicks.

2. **User-Item Matrix:**
   - Organize the data into a user-item matrix, where rows represent users, columns represent items, and the entries contain the users' interactions (e.g., ratings).

3. **Similarity Calculation:**
   - Calculate the similarity between users or items using similarity metrics such as cosine similarity, Pearson correlation, or Jaccard similarity.
   - For user-based collaborative filtering, user similarities are calculated, and for item-based collaborative filtering, item similarities are calculated.

4. **Neighborhood Selection:**
   - For each user or item, select the most similar users or items as the neighborhood.
   - The size of the neighborhood (the number of similar users or items to consider) is an important parameter to control the system's behavior.

5. **Prediction Generation:**
   - Predict the ratings for items that the target user has not yet interacted with by combining the ratings of neighboring users or items.

6. **Recommendation Generation:**
   - Recommend items with the highest predicted ratings to the target user.

**Advantages of Collaborative Filtering using User-Item Interactions:**
- Collaborative filtering is based solely on user interactions and does not require knowledge of item attributes, making it useful for cases where item data is sparse or unavailable.
- It can provide serendipitous recommendations, suggesting items that users may not have discovered on their own.
- Collaborative filtering can be applied in various domains, including e-commerce, music, movie, and content recommendations.

**Limitations of Collaborative Filtering:**
- The cold-start problem: Collaborative filtering struggles to recommend to new users or items with no or limited interaction history.
- It may suffer from sparsity when data is limited or when users have only interacted with a small subset of items.
- Scalability issues can arise with large datasets and an increasing number of users or items.

Here is your task:

1. Write a function that takes in a user id and the dataframe you created before that contains 'user_id', 'title', and 'rating'. The function should return collaborative filtering recommendations for this user based on a user-item interaction matrix. Here are steps you can take:

  A. Create the user-item matrix using Pandas' [pivot_table](https://pandas.pydata.org/docs/reference/api/pandas.pivot_table.html).

  B. Fill missing values with zeros in this matrix.

  C. Calculate user-user similarity matrix using cosine similarity.

  D. Get the array of similarity scores of the target user with all other users from the similarity matrix.

  E. Extract, say the the top 5 most similar users (excluding the target user).

  F. Generate movie recommendations based on the most similar users.

  G. Remove duplicate movies recommendations.

In [7]:
# Collaborative Filtering using User-Item Interactions

def collaborative_filtering_recommendation(user_id, df):
  # Create the user-item matrix, rows=users, columns=items, and entries=ratings
  user_ratings_pivot = df.pivot(index='user_id', columns='movie_id', values='rating')
  # print("Null Cnt:\n", user_ratings_pivot.isnull().values.sum())

  # Fill missing values with 0 (indicating no rating)
  user_ratings_pivot = user_ratings_pivot.fillna(0)
  # print("Null Cnt after fillna:\n", user_ratings_pivot.isnull().values.sum())
  # print("user_ratings_pivot:\n", user_ratings_pivot.head())
  # movie_ratings_pivot = user_ratings_pivot.T
  # similarities = cosine_similarity(movie_ratings_pivot)

  # Calculate user-user similarity matrix using cosine similarity
  similarities = cosine_similarity(user_ratings_pivot)
  cosine_similarity_df = pd.DataFrame(similarities,
                                      index=user_ratings_pivot.index,
                                      columns=user_ratings_pivot.index)

  # Get the similarity scores of the target user with all other users
  cosine_similarity_series = cosine_similarity_df.loc[user_id]

  # Find the top N most similar users (excluding the target user)
  ordered_similarities = cosine_similarity_series.sort_values(ascending=False)

  # Generate movie recommendations based on the most similar users
  print("ordered_similarities:\n", type(ordered_similarities), ordered_similarities.head(6))
  # print(ordered_similarities.iloc[1:6].index)
  similar_user_ids = ordered_similarities.iloc[1:6].index
  similar_user_ratings = df[df['user_id'].isin(similar_user_ids)]
  # print("similar_user_ratings count =", len(similar_user_ratings))
  # print("similar_user_ratings:\n", similar_user_ratings.head(20))

  # Remove duplicates from recommendations
  similar_movies = similar_user_ratings['title'].unique()
  print("Num similar_movies:", similar_movies.shape[0])
  print("Similar (ratings) user_ids:", similar_user_ids)
  return similar_movies

Now, test your recommendations engines! Select a few user ids and generate recommendations using both functions you've written. Are the recommendations similar? Do the recommendations make sense?

In [8]:
# Test the recommendation engines
# user_ratings_pivot = df_movie_ratings.pivot(index='user_id', columns='title', values='rating')
# print(user_ratings_pivot) # Index contains duplicate entries, cannot reshape

print(f"df_movie_ratings shape: {df_movie_ratings.shape}")
print(f"df_movie_ratings columns:\n {df_movie_ratings.columns}")
null_movie_id_cnt = df_movie_ratings['movie_id'].isnull().sum()
null_title_cnt = df_movie_ratings['title'].isnull().sum()
print(f"df_movie_ratings null_movie_id_cnt={null_movie_id_cnt} null_title_cnt={null_title_cnt}")
print('--------------------')

user_ids = [1, 916, 8, 600, 196, 579, 222, 244]
# user_ids = [1, 916]
for id in user_ids:
  print(f"\nTop 5 (collaborative_filtering) similar movies for user_id {id}\n")
  similar_user_movies = collaborative_filtering_recommendation(id, df_movie_ratings)
  print(f"\nTop 5 (collaborative_filtering) similar movies for user_id {id}\n",
        similar_user_movies[0:5])
  print('--------------------')

df_movie_ratings shape: (100000, 5)
df_movie_ratings columns:
 Index(['user_id', 'movie_id', 'title', 'rating', 'genres'], dtype='object')
df_movie_ratings null_movie_id_cnt=0 null_title_cnt=0
--------------------

Top 5 (collaborative_filtering) similar movies for user_id 1

ordered_similarities:
 <class 'pandas.core.series.Series'> user_id
1      1.000000
916    0.569066
864    0.547548
268    0.542077
92     0.540534
435    0.538665
Name: 1, dtype: float64
Num similar_movies: 699
Similar (ratings) user_ids: Index([916, 864, 268, 92, 435], dtype='int64', name='user_id')

Top 5 (collaborative_filtering) similar movies for user_id 1
 ['Toy Story (1995)' 'GoldenEye (1995)' 'Four Rooms (1995)'
 'Get Shorty (1995)' 'Copycat (1995)']
--------------------

Top 5 (collaborative_filtering) similar movies for user_id 916

ordered_similarities:
 <class 'pandas.core.series.Series'> user_id
916    1.000000
94     0.616224
429    0.609329
92     0.593216
276    0.592929
59     0.586565
Name: 916, 

##**Summary:**</br>

Downloaded the movielens data provided.

From the following, see that </br>
 **943 unique users have provided 100000 ratings for 1682 unique movies, spanning 19 (including unknown) genres.**</br>
<code>
ratings shape: (100000, 4)  
movies.shape: (1682, 24)  
ratings - num unique user_id: 943  
ratings - num unique movie_id: 1682  
movies - num unique movie_id: 1682  
num_genres: 19  
</code>
Merged the movies and ratings dataframes and filtered to get a
df_movie_ratings dataframe with the following columns.</br>
['user_id', 'movie_id', 'title', 'rating', 'genres']</br>

Since the genres were binary encoded over separate genre_cols, combined them to a string based 'genres' column, with values such as 'Animation Children Comedy', to allow TF-IDF vectorization.
See that some movies have over 500 ratings, while some have just 1 rating</br>

**Used df_movie_ratings to make both content based recommendations and collaborative filtering recommendations for a given user_id.**</br>

For the function **content_based_recommendation(user_id, df)**, did the following </br>

*   Get the user's rated movies
*   Create a TF-IDF matrix using movie 'genres'
*   Create a dataframe, tfidf_df with a row for each movie, a column for each genre
*   Compute the cosine similarity between movie genres
*   Created df_genre_similarities based on genre similarities of the movies
*   From df_genre_similarities, separated out the movies not seen/rated by the user.
*   With this deduced the top 5 similar_movies for the given user based on 'similarity_score' </br>
  Tried this with a few user_ids. An example below:

<code>
Top 5 (content_based) similar movies for user_id 1 </br>
  
title        similarity_score</br>
Night on Earth (1991)                                       0.731446</br>
Paper, The (1994)                                           0.731446</br>
To Die For (1995)                                           0.731446</br>
Tin Men (1987)                                              0.731446</br>
Adventures of Priscilla, Queen of the Desert, T...          0.731446</br>
</code>

**See that the similarity score is high for the movies recommended.**


For the function, **collaborative_filtering_recommendation(user_id, df)** did the following </br>
*   Create the user_ratings_pivot table with rows=users, columns=movie_id, and entries=ratings
*   Fill missing values with 0 (indicating no rating)
*   Did cosine_similarity on user_ratings_pivot to find similarities for all users
*   Ordered the similarity scores for the given user, to find the top 5 most similar users
*   Got the movies rated by these users, and removed the duplicates.</br>
Tried this on a few user_ids and printed the top 5 movies to recommend.</br>
An example below: </br>

<code>
--------------------</br>
Top 5 (collaborative_filtering) similar movies for user_id 8</br>
Num similar_movies: 220</br>
Similar (ratings) user_ids: Index([**600**, 197, 826, 619, 466]</br>
**Top 5 (collaborative_filtering) similar movies for user_id 8**</br>
 ['Toy Story (1995)' 'GoldenEye (1995)' 'Get Shorty (1995)'
 'Twelve Monkeys (1995)' 'Seven (Se7en) (1995)']</br>
--------------------</br>
Top 5 (collaborative_filtering) similar movies for user_id 600</br>
Num similar_movies: 204</br>
Similar (ratings) user_ids: Index([826, 197, **8**, 619, 217]</br>
**Top 5 (collaborative_filtering) similar movies for user_id 600**</br>
 ['Toy Story (1995)' 'GoldenEye (1995)' 'Get Shorty (1995)'
 'Twelve Monkeys (1995)' 'Seven (Se7en) (1995)']</br>
--------------------</br>
</code>

**Here user_id=600 is in the top 5 users used to pick the movies for user_id=8 and vice-versa.
However, this is not always true.**